Add container.Cache for storing container.PodStatus

yujuhong opened the pull request:
This cache is populated by the generic PLEG, and will be the source of truth
for status of pod/container on the node.

This PR addresses #12540, and should reduce the cpu usage of kubelet and docker.

This PR depends on #17420, #17533, and #17545, and the implementation of the new GetPodStatus function in rkt.
Only the last three commits are new.

yujuhong commented:
/cc @Random-Liu

Random-Liu commented:
Wow~So long a commit list~ :)

yujuhong commented:
> Wow~So long a commit list~ :)

Because merge bot doesn't like our PRs...

yifan-gu commented on pkg/kubelet/pod_workers.go:
```diff
@@ -99,13 +103,20 @@ func newPodWorkers(runtimeCache kubecontainer.RuntimeCache, syncPodFn syncPodFnT
+
 	for newWork := range podUpdates {
 		err := func() (err error) {
+			if p.podCache != nil && !p.podCache.NewerThan(minRuntimeCacheTime) {
```
is anyone using the podCache here?

yujuhong replied:
Nope, now it's here to ensure the workers don't sync prematurely. They'd wait until the cache is up-to-date (newer than the end of last sync).

We could pass this to syncPod here if we want. This PR needs some more work after the rest of the PodStatus PRs are merged.

yujuhong committed:
Always overwrite items in kubelet's work queue

This allows kubelet to change the next sync time based on the last result.

yujuhong commented:
- Changed the Get(), Set() method to accept pod UID (instead of `*kubecontiner.Pod`)
- Added the callback mechanism to wake up the pod workers
- Changed the relist period of PLEG from 3s to 1s now that it'd affect the latency of pod status generation

@yifan-gu @Random-Liu PTAL, thanks!

yujuhong commented:
Addressed the comments. Ready for review (again). Thanks!

yujuhong commented:
@k8s-bot unit test this

yujuhong commented:
Addressed comments and rebased.

yujuhong commented:
@k8s-bot unit test this

Random-Liu commented:
LGTM, rebase please. :)

yujuhong commented:
@Random-Liu thanks for reviewing. I still need to fix integration tests which use the FakeDockerClient.

In the meantime, @kubernetes/goog-node, @yifan-gu any feedback is welcome.

yifan-gu commented:
Will review today.

yujuhong commented:
Fixed the integration test, rebased, and corrected the typo.

timstclair commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,187 @@
+// and the blocking GetNewerThan() method. The component responsible for
+// populating the cache is expected to call Set() to set the status to empty to
+// reclaim the cache entries.
+type Cache interface {
```
How does this relate to `status.Manager`? Which will be the source of truth? (e.g. for probe workers).

yujuhong replied:
`status.Manger` is the source of truth of `api.PodStatus`, while `Cache` the source of truth of the internal `PodStatus`. It depends on whether or not prober relies on information only available in `api.PodStatus`.

timstclair replied:
Ah, sorry, I missed that this was a different `PodStatus` type. That makes more sense.

The probe worker needs the ContainerID, start time, and whether the container is terminated. All that is available in the `kubecontainer.PodStatus`, so will this be the prefered method for retrieving that?

timstclair commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,187 @@
+type Cache interface {
+	Get(types.UID) (*PodStatus, error)
+	Set(types.UID, *PodStatus, error, time.Time)
+	// GetNewerThan is a blocking call that would return the status
```
s/would return/only returns/  ?

yujuhong replied:
Done

timstclair commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,187 @@
+func (c *cache) notify() {
+	for id, r := range c.callbacks {
+		if c.timestamp.After(r.time) {
+			nonblockingSend(r.ch, c.get(id))
```
Consider closing the channel before deleting it? That will prevent a routine from getting stuck waiting for a callback when the channel is deleted.

yujuhong replied:
This doesn't really delete the channel, but just delete the callback record in the map. Since `GetNewerThan()` always creates a new channel, it should not get stuck as long as there is a notification inflight. I'll close the channel just in case, but there is nothing the goroutine can do if the channel gets closed without any preceding notification...

timstclair commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,187 @@
+func (c *cache) addCallback(id types.UID, timestamp time.Time, ch chan *data) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.callbacks[id] = callbackRecord{time: timestamp, ch: ch}
```
Consider checking whether `c.callbacks[id]` is already set, and logging an error in that case.

yujuhong replied:
Good point. I extended the code to allow multiple callers of GetPodIfNewer for the same pod id, and makes the `callbacks[id]` maps to a list.

yifan-gu commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,187 @@
+		return d.status, d.err
+	}
+	ch := make(chan *data, 1)
+	c.addCallback(id, minTime, ch)
```
There is a window between getIfNewerThan and addCallback.
Imagine the cache is updated in that window, the listener on the channel will not be notified.

Random-Liu replied:
If that happens, It will be notified in the next relist(). However, it's better to put these into one critical block.

yujuhong replied:
It'll eventually be notified due to relisting, but yes, the code should not rely on that. Moving the `getIfNewerThan()` check to `addCallback` so we always check while holding the lock.

yifan-gu commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,187 @@
+
+	if !ok {
+		// Status is not cached.
+		if c.timestamp != nil && c.timestamp.After(minTime) {
```
I still think we can avoid the nil pointer by using time.Time instead of *time.Time in the cache struct.

We can use IsZero() bool to test if the timestamp is set.

yujuhong replied:
The first time a pod worker syncs, it will also have the last sync time as zero :-(

yifan-gu commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,187 @@
+	// it is ready to serve the cached statuses.
+	timestamp *time.Time
+	// Map that stores the callback records.
+	callbacks map[types.UID]callbackRecord
```
callbacks is not the best name here, as callbacks are mostly executable function. These are just channels with a timestamp, so maybe something like watchers, notifiers

yujuhong replied:
I'm not good with names, but will try coming up with better ones.

yifan-gu commented on pkg/kubelet/pod_workers.go:
```diff
@@ -99,13 +101,25 @@ func newPodWorkers(runtimeCache kubecontainer.RuntimeCache, syncPodFn syncPodFnT
+				// Time. This ensures the worker doesn't start syncing until
+				// after the cache is at least newer than the finished time of
+				// the previous sync.
+				p.podCache.GetNewerThan(podID, minRuntimeCacheTime)
```
??

yujuhong replied:
This just blocks the worker until we are ready to sync. We can actually get the PodStatus and pass it to syncPod, but I want to avoid extra churns/conflicts so they will be addressed later.

yifan-gu commented on pkg/kubelet/pod_workers.go:
```diff
@@ -99,13 +101,25 @@ func newPodWorkers(runtimeCache kubecontainer.RuntimeCache, syncPodFn syncPodFnT
 	for newWork := range podUpdates {
 		err := func() (err error) {
+			podID := newWork.pod.UID
+			if p.podCache != nil {
```
question here, when podCache will be nil?

Random-Liu replied:
When pod cache is disabled, probably for test or some other reason.

yujuhong replied:
I only used podCache if the container runtime is docker. This way, we were not blocked on  changes in rkt (which may no longer be needed). It also provides a simple knob to turn off podCache while testing.

yifan-gu replied:
@yujuhong @Random-Liu OK, can we add a todo here to remind us to clean this up later?

yifan-gu replied:
btw, why the cache is disabled for rkt?

Random-Liu replied:
I think it's because when this PR was initially created (long long ago :P), rkt had no support for kubelet internal pod status yet.

yujuhong commented:
@timothysc, this PR should fix/alleviate the burst of cpu (~2 cores, kubelet and docker each) that you see in the steady cluster (when there is no work). It doesn't address the case of batch creation and/or deletion. For that, we could consider simply rate-limiting the docker requests.

yifan-gu commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,204 @@
+
+// getIfNewerThan returns the data it is newer than the given time.
+// Otherwise, it returns nil. The caller should acquire the lock.
+func (c *cache) getIfNewerThan(id types.UID, minTime time.Time) *data {
```
@yujuhong I found there are 3 things to be returned:

- nil
- valid pod status
- empty pod status (only has pod id)

Can you explain when one of these is returned? Why would there be empty pod status?

yujuhong replied:
- nil: we don't have any status for this pod that's newer than |minTime|
- valid pod status: the pod has visible containers on the node, and the status is stored in the cache. The status is newer than |minTime|
- empty pod status: the pod has no visible containers on the node, therefore it's not cached. We detected this at a time that's newer than |minTime|

I made a conscious decision to not store empty pod status so that we don't have to garbage collect entries explicitly. An empty pod status here just means that there is no visible containers associated with the pod, so the status is valid.

I can expand the comment to explain this more clearly.

yujuhong replied:
I think I miss one case in the PLEG where the cache entries should be garbage collected. Will fix that and also consider adding an explicit `Delete()` method for the cache, if that make things more clear.

yifan-gu commented on pkg/kubelet/pleg/generic_test.go:
```diff
@@ -146,3 +149,96 @@ func TestRelisting(t *testing.T) {
+		assert.Equal(t, c.status, actualStatus, testStr)
+		assert.Equal(t, c.error, actualErr, testStr)
+		if c.event != nil {
+			assert.Contains(t, actualEvents, c.event, testStr)
```
@yujuhong There should be only one expected event right? Can we test it outside of the loop?

yifan-gu replied:
Can we test the exactly events instead of `Contains`?

yujuhong replied:
Done.

yifan-gu commented on pkg/kubelet/kubelet.go:
```diff
@@ -120,7 +120,7 @@ const (
 	// Generic PLEG relies on relisting for discovering container events.
 	// The period directly affects the response time of kubelet.
-	plegRelistPeriod = time.Second * 3
+	plegRelistPeriod = time.Second * 1
```
Just curious, what guides us for a proper relisting frequency ?
 Add a line note

yujuhong replied:
Will do. It's a tradeoff between resource usage and how responsive kubelet will appear.

- We run `docker ps -a` every period to detect the changes, and inspect the changed containers if they exist. This operation could become more expensive as the number of containers grow. Shorter period would cause higher cpu consumption.
- After receiving a new pod, kubelet starts the containers and wait ~1 relist period to detect the changes and report the running status. Increasing the period will cause longer delay.

I'll comment more in the code.

yifan-gu replied:
@yujuhong Sorry, "add a line note" is not my word, it's the github text I accidently copied when moving my comments from commits to here :)

But more docs are welcome. Did you do any experiments so you changed the default period?

yifan-gu commented on pkg/kubelet/kubelet.go:
```diff
@@ -390,7 +391,8 @@ func NewMainKubelet(
 		klet.containerRuntime = rktRuntime
-		klet.pleg = pleg.NewGenericPLEG(klet.containerRuntime, plegChannelCapacity, plegRelistPeriod)
+		klet.podCache = nil
+		klet.pleg = pleg.NewGenericPLEG(klet.containerRuntime, plegChannelCapacity, plegRelistPeriod, klet.podCache)
```
This line is the same as line 373, can you move it outside of the switch-case ?

yifan-gu commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,204 @@
+	if _, ok := c.subscribers[id]; !ok {
+		c.subscribers[id] = []*subRecord{}
+	}
+	c.subscribers[id] = append(c.subscribers[id], &subRecord{time: timestamp, ch: ch})
```
why will there be concurrent multiple subscribers for a pod?

yujuhong replied:
For now, we don't have multiple subscribers. However, it seems dumb to have a GetPodIfNewer() method in a cache that can not support multiple callers for a single pod.

yifan-gu replied:
ok

yifan-gu commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,204 @@
+	}
+}
+
+func (c *cache) subscribe(id types.UID, timestamp time.Time, ch chan *data) {
```
can we return a channel here, we can make sure the channel has buffer size >=1 so that we can remove `nonblockingSend`

yujuhong replied:
Done.

yifan-gu commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+func (c *cache) GetNewerThan(id types.UID, minTime time.Time) (*PodStatus, error) {
+	ch := make(chan *data, 1)
```
ditto, let subscribe to return a channel is better imo

yujuhong replied:
Done.

yifan-gu commented:
@yujuhong Besides the comments, my question about the cache:
what will we do when we get miss? do we want to make another try and populate the cache or not?

yujuhong commented:
> @yujuhong Besides the comments, my question about the cache:
> what will we do when we get miss? do we want to make another try and populate the cache or not?

The cache should container all visible containers on the node. In the case where the pod is not in the internal storage of the cache, the cache will return an empty pod status, i.e., there is no visible container (running or dead) on the node associated with this pod.

If, for some reason, PLEG cannot inspect the container, it will store an error in the cache. PLEG will try to inspect again in the next relist.

yujuhong commented:
Thanks for the review. I need to fix some bug in the logic first. Expect some changes to the PR later.

yujuhong committed:
PLEG should report events if a container is removed

Currently, pleg would report a event if a container transitions from running to
exited between relisting. However, if would not report any event if a container
gets stopped and removed between relisting. This event will eventually be
handled when the pod syncs periodically, but this is undesirable. This change
ensures that we detect all such events.

yujuhong committed:
Add container.Cache for storing container.PodStatus

This cache will be used to stores the PodStatus of all pods/containers
visible on the node. This will elimiate the need for pod workers to query the
container runtime directly.

yujuhong commented:
@yifan-gu @Random-Liu 
I forked two trivial testing PRs #19392, #19393 to reduce the noise of this review.
After #19369 is finalized, I'll resume working on this PR, and may further split it into two or more PRs to help with the review process.

yujuhong committed:
Support populating the runtime cache in PLEG

This changes does not turn on this feature (cache) for kubelet.

yujuhong commented:
The following three PRs are new, the rest are dependencies on to-be-merged PRs.
* Overwrite items in kubelet's work queue: b02673d7ca5de62e274a24ceeec5927102864160: Same as before. No new change.
* Add a container.Cache for storing container.PodStatus: 844af2a2268a9d450ebed48e18b8b41809bdfaed: Addressed the comments. PTAL again.
* Support populating the runtime cache in PLEG: 2486e7e80eaa137646ce4137124fb699e68c87a7: Mostly rewritten based on #19369. PTAL.

I'll submit a separate PR to actually instruct kubelet to use the cache.

yujuhong commented:
I pushed again just change some minor things in cache.go. Sorry if this caused some troubles for reviewing.

yifan-gu commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,204 @@
+		return ch
+	}
+	// Add the subscription record.
+	if _, ok := c.subscribers[id]; !ok {
```
It seems the if clause can be removed. https://play.golang.org/p/CsbkpQmmZm

yujuhong replied:
done.

yifan-gu commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+// SetTimestamp modifies the global timestamp of the cache.
+func (c *cache) SetGlobalTimestamp(timestamp time.Time) {
```
This function not only sets the global timestamps, it also notifies the subscribers, so `SetGlobalTimestamp` doesn't sound like a proper name as it tells the user nothing about its side effect. But I don't have a good suggestion. But how bad is splitting this into two functions?

yujuhong replied:
changed to `UpdateTime` to be less specific.

yifan-gu commented on pkg/kubelet/container/cache_test.go:
```diff
@@ -0,0 +1,210 @@
+		timestamp time.Time
+		expected  bool
+	}{
+		{
```
nits: As we are comparing the `cacheTime` and `modified` with the `timestamp`. Would it be more straightforward if we let the `timestamp` be constant, and control the `modified` and `cacheTime` ?

yujuhong replied:
done.

yifan-gu commented on pkg/kubelet/kubelet.go:
```diff
@@ -387,7 +387,7 @@ func NewMainKubelet(
 		}
 		klet.containerRuntime = rktRuntime
-		klet.pleg = pleg.NewGenericPLEG(klet.containerRuntime, plegChannelCapacity, plegRelistPeriod)
+		klet.pleg = pleg.NewGenericPLEG(klet.containerRuntime, plegChannelCapacity, plegRelistPeriod, nil)
```
[This](https://github.com/kubernetes/kubernetes/pull/17700#discussion_r49015593) comment get lost
> This line is the same as line 373, can you move it outside of the switch-case ?

yujuhong replied:
Nope, I split the PR out to #19436 and addressed the comment there.

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -116,40 +158,191 @@ func (g *GenericPLEG) relist() {
 	// Get all the pods.
-	pods, err := g.runtime.GetPods(true)
+	podList, err := g.runtime.GetPods(true)
+
```
unnecessary empty line

yujuhong replied:
removed

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -49,24 +49,54 @@ type GenericPLEG struct {
-type containerInfo struct {
-	podID types.UID
-	state kubecontainer.ContainerState
+// plegContainerState has an one-to-one mapping to the
```
s/an/a

yujuhong replied:
done

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -49,24 +49,54 @@ type GenericPLEG struct {
-	podID types.UID
-	state kubecontainer.ContainerState
+// plegContainerState has an one-to-one mapping to the
+// kubecontainer.ContainerState except for the Non-existent state. This state
```
s/Non/non

yujuhong replied:
done

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -116,40 +158,191 @@ func (g *GenericPLEG) relist() {
+}
+
+func (g *GenericPLEG) updateCache(pod *kubecontainer.Pod) error {
+	timestamp := time.Now()
```
remove this?

yujuhong replied:
remove the tiemstamp? The timestamp needs to be *before* we query the container runtime to be safe.

yifan-gu replied:
you are right

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -116,40 +158,191 @@ func (g *GenericPLEG) relist() {
+		// Update the cache timestamp.  This needs to happen *after*
+		// all pods have been properly updated in the cache.
+		g.cache.SetGlobalTimestamp(timestamp)
+	} else {
```
just return above and remove the else here?

yujuhong replied:
I think it's more clear to keep the if-else blocks since `return` the blocks are quite big -- people may miss the return statement.

yifan-gu replied:
@yujuhong hmm, how about move the else part to the if part:
```go
if !g.cacheEnabled() {
    g.podRecords.updateAll()
    for ... {
    }
    return
}

// Cache enabled.
for pid, events := range eventsByPodID {
    ....
}
g.cache.UpdateTime(timestamp)

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -116,40 +158,191 @@ func (g *GenericPLEG) relist() {
+			// parallelize if needed.
+			if err := g.updateCache(pod); err != nil {
+				glog.Errorf("PLEG: Ignoring events for pod %s/%s: %v", pod.Name, pod.Namespace, err)
+			} else {
```
can you use continue and remove the else here?

yujuhong replied:
done

yifan-gu commented on pkg/kubelet/pleg/generic_test.go:
```diff
@@ -146,3 +148,179 @@ func TestRelisting(t *testing.T) {
+		assert.Equal(t, c.status, actualStatus, testStr)
+		assert.Equal(t, c.error, actualErr, testStr)
+		if c.event != nil {
+			assert.Contains(t, actualEvents, c.event, testStr)
```
Can we test the exactly events instead of Contains?

yujuhong replied:
Done

yifan-gu commented:
Reviewed, most comments are small nits. No big issues.

yujuhong commented:
> s/reilst/relist

Done.

yujuhong commented:
> Reviewed, most comments are small nits. No big issues. 
> Besides this, I would suggest splitting the cache implementation and the cache integration of pleg into two PRs next time :) (as the comments on commits are easy to get lost)

I agree commits are not good for commenting. On the other hand, PR merging is so slow that any extra PR will cause significant time (rerunning the tests, etc), hope this will get better soon :-(

Thanks for reviewing. Addressed all comments. Will squash once I get the LGTM.

yujuhong commented:
Out of the three PRs that this PR depends on, two of them were manually merged. The remaining one (#19369) is still waiting in the queue. I'll rebase after the PR is merged and/or review is completed.

yujuhong commented:
@yifan-gu PTAL, thanks!

yifan-gu commented:
@yujuhong reviewing.

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -116,40 +158,190 @@ func (g *GenericPLEG) relist() {
+		// Update the cache timestamp.  This needs to happen *after*
+		// all pods have been properly updated in the cache.
+		g.cache.UpdateTime(timestamp)
+	} else {
```
conversation is missed here:

>yujuhong added a note a day ago
I think it's more clear to keep the if-else blocks since return the blocks are quite big -- people may miss the return statement.
 @yifan-gu  

>yifan-gu added a note 20 hours ago
@yujuhong hmm, how about moving the else part to the if part:

if !g.cacheEnabled() {
    g.podRecords.updateAll()
    for ... {
    }
    return
}

// Cache enabled.
for pid, events := range eventsByPodID {
    ....
}
g.cache.UpdateTime(timestamp)

yujuhong replied:
Done.

yifan-gu commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,202 @@
+
+//  UpdateTime modifies the global timestamp of the cache and notify
+//  subscribers if needed.
+func (c *cache) UpdateTime(timestamp time.Time) {
```
Hmm, I still feel a little bit raw about this function name is not reflecting that the subscribers are being notified.

yujuhong replied:
Notifying the subscriber is part of the =internal implementation of this cache. The interface doesn't mention anything about subscriptions at all. I don't think we should expose the details through the function name.

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -116,40 +158,190 @@ func (g *GenericPLEG) relist() {
+		oldPod := g.podRecords.getOld(pod.ID)
+
+		// Process all currently visible containers in the pod.
+		for _, container := range pod.Containers {
```
I feel it's cleaner to move the two for loop to a function which takes (oldPod, newPod)

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -116,40 +158,190 @@ func (g *GenericPLEG) relist() {
+		}
+	}
+	// Process all pods that are no longer visible.
+	for pid := range g.podRecords {
```
also this loop

yifan-gu commented on pkg/kubelet/container/cache.go:
```diff
@@ -0,0 +1,202 @@
+	}
+	// The pod status is not ready.
+	return nil
+}
```
How about
```go
if !ok && c.timestamp != nil && c.timestamp.After(minTime) {
    return makeDefaultData(id)
}
if ok && d.modified.After(minTime) || c.timestamp.After(minTime) {
    return d
}
return nil
```

yujuhong replied:
Done.

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -116,40 +159,194 @@ func (g *GenericPLEG) relist() {
+	} else if newPod != nil {
+		pid = newPod.ID
+	} else {
+		// Both pods are nil; do nothing.
```
How this could happen?

yujuhong replied:
should not happen. removed.

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -116,40 +159,194 @@ func (g *GenericPLEG) relist() {
+			}
+		}
+		return
+	}
```
Can't we delete line189-line195 and resue line 222-line 226?

yujuhong replied:
done

yifan-gu commented:
@yujuhong Thank you for the hard work! LGTM!

yujuhong commented:
@yifan-gu, thanks for the thorough reviews and good suggestions!

Squashed the commits. Still waiting for #19369 to land first.

timothysc commented:
@yujuhong could you please ping me when this is in.

yujuhong commented:
Fixed the unit test. Re-applying the lgtm label. 

@timothysc, there is a followup PR to start using this cache. I'll let you know when that PR gets in as well.
